### ISSUE TITLE
quicksort example stack overflows on large input 

### ISSUE THREAD

**Issue by m4b** [nikomatsakis/rayon#376]
I was hoping I could be super lazy and copy paste the parallel quicksort, but alas, it seems that for very large inputs this can stack overflow, as it recurses.

On that note are there any plans on introducing an efficient parallel sort into the lib or is that outside the scope of rayon ? (I frequently want this, but I don't have the time to mess around with it and was hoping there was a solution already)

**Comment by stjepang** [comment]
Hey, just wanna say that I would be interested in implementing a fast parallel sort. :)
Since I've recently implemented `slice::sort` and `slice::sort_unstable`, perhaps that would be a good follow-up problem to solve?

By the way, C++ has got a bunch of [parallel algorithms](http://en.cppreference.com/w/cpp/experimental/parallelism) (click on `[Expand]`) - are there any important functions besides sorting that Rayon is missing? Can we make a list of them?

**Comment by cuviper** [comment]
In the best case it should only have O(log n) recursion depth, but on an already-sorted array I think it will have linear depth, since the current pivot index is just the last element.  See wikipedia on [choice of pivot](https://en.wikipedia.org/wiki/Quicksort#Choice_of_pivot) -- so we could do better there.

I've thought about adding sorting methods to `ParallelSliceMut`.  I think unstable sort wouldn't be too bad, as you can just partition in-place to split jobs and then choose any sequential algorithm you like.  I'm not sure of a good way to do stable partitioning though.  If you want to tackle this @stjepang, feel free!

The C++ comparison would be an interesting task... maybe we should start a wiki page for that like we did for basic `Iterator` parity.

**Comment by m4b** [comment]
Ugh I will be so happy if this exports a nice sorting algo I can just use without any effort. Will <3 you all forever 

For reference I was getting stack overflows sorting with example parallel quicksort on 150million element nonsorted vector totallling about 50GB. Profiling shows it takes about 7% of process time so hoping parallel version could trim that down significantly

**Comment by stjepang** [comment]
All right, I've just created this work-in-progress project: https://github.com/stjepang/par-sort
As soon as it gets finished and demonstrates satisfying benchmark numbers I'll submit a PR moving it into Rayon.

It includes two algorithms:

1. Quicksort (unstable sort): functions  `sort_unstable{,_by,_by_key}`. Recursive steps are performed in parallel, but partitioning is still sequential.
2. Mergesort (stable sort): functions `sort{,_by,_by_key}`. It's still totally sequential so don't bother with it. :)

@m4b Unstable sorting should already be pretty performant. If you want, you can give it a shot and let me know how it works out. :)

On my dual-core machine (2 physical, 4 logical cores) the `simple` example (sorting 100 million `u64`s) takes:

* 8.6 sec using `v.sort()`
* 4.8 sec using `v.sort_unstable()`
* 2.8 sec using `v.par_sort_unstable()`

@cuviper I have a question regarding the API we'd like to expose in Rayon. In order to parallelize partitioning in quicksort, we must share the pivot among multiple threads. This requires a new type bound `T: Sync`. Is this additional type bound too restrictive - what do you think?

With the help of specialization we could use parallel partitioning when `T: Sync` and sequential partitioning otherwise. But specialization is available on nightly Rust only.

The current API I've sketched out can be found [here](https://github.com/stjepang/par-sort/blob/master/src/lib.rs#L8).

**Comment by cuviper** [comment]
> In order to parallelize partitioning in quicksort, we must share the pivot among multiple threads. This requires a new type bound T: Sync. Is this additional type bound too restrictive - what do you think?

It would be nicer not to require that, but maybe you can convince me with some performance numbers.  You're already doing pretty good for dual-core results, as hyperthreading often seems not to add much in Rayon benchmarks.

I wasn't even aware that parallel partitioning is possible -- can you sketch out how that would work?

A minor note on implementation - I do expect this to go directly on `ParallelSliceMut` when you merge into rayon, unless you have some reason why it should be a new trait.  The style there is to use `as_parallel_slice_mut()` so methods can be directly implemented by default, instead of in a separate `impl` block.

I also kind of expected these to just use the standard library's `sort()` once you get down to sequential work -- at least in the stable case for now, and eventually for `sort_unstable()` too when we bump our rust requirement.  Any reason you're not doing that?

(I only briefly skimmed the implementation, so sorry if I missed that explanation already.)

**Comment by stjepang** [comment]
> It would be nicer not to require that, but maybe you can convince me with some performance numbers.

I don't have conclusive numbers yet, but my guess (based on some tweaking and benchmarking) is that parallel partitioning would improve the running time by 10-15%.

Can we add `nightly` feature to Rayon? Crossbeam has it defined [here](https://github.com/crossbeam-rs/crossbeam/blob/master/Cargo.toml#L13) and uses unstable language features when it's enabled. I like that. :)

> I wasn't even aware that parallel partitioning is possible -- can you sketch out how that would work?

An explanation can be found [here](http://www.lsi.upc.edu/~lfrias/research/parpar/wea08.pdf).

The short story is: Suppose you have chosen a pivot and have N threads. Split the array into blocks of equal size. Now each thread performs the following algorithm:

First it atomically takes one block from the left end and one block from the right end. Then it partitions the elements within these blocks until one of those is filled up. If the left block gets filled first, it atomically takes another block from the left end, and viceversa if the right block gets filled first. Then it continues partitioning and repeats this procedure until no more blocks are left.

At the end at most N blocks will be unfinished (because we have N threads). Now you just have to partition the elements among these remaining blocks - this part can be done sequentially or in parallel (but it may not be worth the implementation effort).

> I do expect this to go directly on `ParallelSliceMut` when you merge into rayon, unless you have some reason why it should be a new trait.

You're right , there is no reason for a new trait - it'll go into `ParallelSliceMut`.

> I also kind of expected these to just use the standard library's `sort()` once you get down to sequential work -- at least in the stable case for now, and eventually for `sort_unstable()` too when we bump our rust requirement. Any reason you're not doing that?

Regarding stable sort, soon I'm going to push major changes to the way the work is initially split among threads. This will improve performance on random slices as well as on slices containing long ascending/descending runs.

Regarding unstable sort, it's best to parallelize recursive quicksort calls and that means changing the implementation of quicksort. A few other tweaks are necessary like limiting slice length for partial insertion sort etc.

Unfortunately, we have to change the guts of these algorithms in order to get all the performance we can. Drawing a line between slice division and sequential work in the form of `v.sort()` or `v.sort_unstable()` would incur significant performance penalties.

Anyways, finishing the implementation and thoroughly testing it will take at least several weeks.
Btw, some good news: with the new changes `par_sort` is faster than `__gnu_parallel::sort` and `tbb::parallel_sort` in C++, at least with random arrays of integers and on my dual-core machine.

**Comment by cuviper** [comment]
> parallel partitioning would improve the running time by 10-15%.

Hmm, definitely beneficial -- thanks for the reference and TL;DR summary. :)  IMO it's better saved for specialization though.  By keeping the interface more general and only requiring `T: Send` up front, we can still support parallel sorting on stuff like `Cell` contents that can't be `Sync`.

Anyway, note that it's only the initial partition that's fully sequential.  Further partitioning the subslices will start working in parallel to each other.

I remain skeptical about the need to avoid `slice::sort()`, but I'll wait and see when you're finished.

**Comment by m4b** [comment]
Tried out the git par_sort repo. Unstable reduces a 22second sort to about a second for my usecase :heart_eyes:

**Comment by stjepang** [comment]
@m4b Wow, that's a huge speedup! :)

May I just ask how many cores does your machine have?
Also, what was the original sort that took 22 seconds - is it the standard `slice::sort` or the quicksort from Rayon's examples?

**Comment by m4b** [comment]
Apologies, it was actually approximately 200seconds to 5-6 seconds ! That number seemed ludicrous but I just double checked :) really incredible work @stjepang!

P.S. I'm not validating the results, so I assume it's actually sorting 😅 Also unstable is ok for my usecase because there aren't duplicate entries

**Comment by cuviper** [comment]
30-40x faster -- gee, is that all? 😜

**Comment by Basilisk46** [comment]
here’s a concise reply https://stackdev.space/questions/quicksort-a-closer-look-at-two-key
